crates/metal-llvm/src/codegen.rs

```diff
@@ -102,96 +115,38 @@ impl CodeGen {
                 let m = m1.as_str();
                 name += m;
             }
-            let n = name + ty.ident.inner.len().to_string().as_str() + ty.ident.inner;
+            let n = name
+                + definition.signature.name.len().to_string().as_str()
+                + definition.signature.name.as_str();
             n
         } else {
-            // this only applies to
-            // main.mt, the entry point of a program.
-            ty.ident.inner.to_string()
+            definition.signature.name.clone()
         };
 
-        unsafe {
-            // TODO: variadic argument support
-            // like `*args`
-            let c_fun_name = CString::new(fun_name).unwrap();
+        let linkage = match definition.signature.vis {
+            metal_mir::types::visibility::Visibility::Parcel => LLVMLinkage::LLVMInternalLinkage,
+            metal_mir::types::visibility::Visibility::Private => LLVMLinkage::LLVMInternalLinkage,
+            metal_mir::types::visibility::Visibility::Public => LLVMLinkage::LLVMExternalLinkage,
+        };
 
+        unsafe {
             let function = LLVMAddFunction(
```

Comment:
same pitfall, `fun_name` is not a CString and the cast will be unnecessary once it is